### DISCUSSION TITLE
How to decode Structures with optional fields

### DISCUSSION THREAD

**Discussion by May1592** [gopcua/opcua#697] [Q&A]
Hi @magiconair @kung-foo ,
I don't see support for decoding structure with optional fields. In below screenshot , it's a custom structure with optional fields and while we decode this we should skip these struct fields since Encoding mask is zero.

![image](https://github.com/gopcua/opcua/assets/41743235/2e74ce36-30cd-4183-ba40-5b567a5b3f98)

Please suggest.

**Comment by magiconair** [comment]
We don't have direct support for optional fields. You need to define your structure with the correct `EncodingMask`. Check similar structures in the `ua` package. I'm open for suggestions on what good support for optional fields should look like.

**Reply by May1592** [reply]
Based On EncodingMask, we will know which optional field is available and which is not. Optional Field which is not available will be null.
Hence in decoding logic we need to take care if Optional field is present or null.
@magiconair your thought on how to build logic for decoding optional field?
https://reference.opcfoundation.org/Core/Part6/v104/docs/5.2.7#_Ref401507564
